File: problems/901/solution.js

var StockSpanner = function() {
    this.index = -1;
    this.stack = [];
};

/** 
 * @param {number} price
 * @return {number}
 */
StockSpanner.prototype.next = function(price) {
    let ans = 0;
    
    this.index++;
    
    while (this.stack.length > 0 && this.stack[this.stack.length - 1][1] <= price) {
        this.stack.pop();
    }
    
    if (this.stack.length > 0) {
        ans = this.index - this.stack[this.stack.length - 1][0];
    } else {
        ans = this.index + 1;
    }
    
    this.stack.push([this.index, price]);
    
    return ans;
};

/** 
 * Your StockSpanner object will be instantiated and called as such:
 * var obj = new StockSpanner()
 * var param_1 = obj.next(price)
 */

File: problems/901/solution.go
package main
type StockSpanner struct {
    idx int
    stack [][]int
}


func Constructor() StockSpanner {
    return StockSpanner{-1, [][]int{}}
}


func (this *StockSpanner) Next(price int) int {
    ans := 0
    
    this.idx++
    
    for len(this.stack) > 0 && this.stack[len(this.stack) - 1][1] <= price {
        this.stack = this.stack[:len(this.stack) - 1]
    }
    
    if len(this.stack) > 0 {
        ans = this.idx - this.stack[len(this.stack) - 1][0]
    } else {
        ans = this.idx + 1
    }
    
    this.stack = append(this.stack, []int{this.idx, price})
    
    return ans
}


/**
 * Your StockSpanner object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Next(price);
 */